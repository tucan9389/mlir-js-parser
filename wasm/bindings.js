// Minimal JS wrapper to call the C API from the Emscripten-generated module.
// Assumes the Emscripten build exports the function `mlir_parse_to_string`.

export async function createParserModule(moduleFactory) {
  // moduleFactory is the Emscripten Module factory (e.g., import from generated .js)
  const maybeModule = await moduleFactory();
  const Module = (maybeModule && typeof maybeModule.then === 'function')
    ? await maybeModule
    : maybeModule;
  if (Module && Module.ready && typeof Module.ready.then === 'function') {
    await Module.ready;
  }

  function parseMlir(text) {
    const c_mlir = Module.cwrap('mlir_parse_to_string', 'number', ['string', 'number', 'number', 'number', 'number']);

    // Allocate output and error buffers (start with 64KB each; grow on demand)
    let outCap = 64 * 1024;
    let errCap = 8 * 1024;
    let outPtr = Module._malloc(outCap);
    let errPtr = Module._malloc(errCap);

    try {
      let rc = c_mlir(text, outPtr, outCap, errPtr, errCap);
      if (rc === 0) {
        const outStr = Module.UTF8ToString(outPtr);
        return { ok: true, module: outStr };
      }
      if (rc < 0) {
        const need = -rc;
        Module._free(outPtr);
        outPtr = Module._malloc(need);
        outCap = need;
        rc = c_mlir(text, outPtr, outCap, errPtr, errCap);
        if (rc === 0) {
          const outStr = Module.UTF8ToString(outPtr);
          return { ok: true, module: outStr };
        }
      }
      const errStr = Module.UTF8ToString(errPtr) || 'parse failed';
      return { ok: false, error: errStr };
    } finally {
      Module._free(outPtr);
      Module._free(errPtr);
    }
  }

  return { Module, parseMlir };
}

// UTF8ToString is provided by Emscripten EXPORTED_RUNTIME_METHODS
