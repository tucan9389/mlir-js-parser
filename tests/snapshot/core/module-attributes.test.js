import { describe, it, expect } from 'vitest';
import { createParserModule } from '../../../wasm/bindings.js';

let ModuleFactory = null;
try {
  const mod = await import('../../../wasm/mlir_parser.js');
  ModuleFactory = mod.default ?? mod;
} catch (e) {
  // swallow, will skip when not present
}

// Snapshot: ensure stable AST for a module with dict attributes

describe('Snapshot: module with attributes', () => {
  it('matches snapshot for attributes dictionary', async () => {
    if (!ModuleFactory) return; // skip if wasm artifact not available
    const { parseMlirJson } = await createParserModule(ModuleFactory);
    const input = 'module attributes { test.meta = { a = 1 : i32, msg = "hi" } } {}';
    const res = parseMlirJson(input);

    expect(res.ok).toBe(true);
    // Ensure keys we care about exist before snapshotting
    expect(res.json).toBeTypeOf('object');
    expect(res.json.name).toBe('builtin.module');
    // Snapshot entire JSON to catch structural changes early
    expect(res.json).toMatchSnapshot();
  });
});
