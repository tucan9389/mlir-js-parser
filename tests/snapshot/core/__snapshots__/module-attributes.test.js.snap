// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot: module with attributes > matches snapshot for attributes dictionary 1`] = `
{
  "attributes": {
    "test.meta": {
      "a": "1",
      "msg": "hi",
    },
  },
  "name": "builtin.module",
  "operands": [],
  "regions": [
    {
      "blocks": [
        {
          "arguments": [],
          "operations": [],
        },
      ],
    },
  ],
  "results": [],
}
`;
